### PR TITLE
Optimize database utilization visualization

### LINKED ISSUE #258: [8] Database optimization
This issue should fix the long time it takes for live data to be requested from the database. 

As a user, I want a quick loading time when clicking through the graph so that I cannot get annoyed by non-responsive visualization.

### PR DESCRIPTION
**Pull request by MaxVanDeursen**
In this PR, the way of querying the database receives an overhaul.
The queries for each individual edge are removed and are instead placed
in one big query, which speeds up the process drastically. There are now
new methods in place which can correctly handle the new format and
convert it to the same format as before this commit.

Fixes #258

### CODE REVIEW

File: Tribler/community/multichain/community.py [hunk 1]
```diff
@@ -293,34 +293,97 @@ def get_statistics(self, public_key=None):
...
+        If there is not yet an entry for the public key pair, it is automatically added to the edges dictionary. If
+        an edge is already recorded, the given amount is added to the total
+
+        :param from_pk: the public key of the node where the edge originates from
+        :param to_pk: the public key of the node where the edge arrives
+        :param edges: the dictionary of edges
+        :param amount: the amount of data transferred over this edge
+        """
+        if from_pk in edges:
+            edges[from_pk][to_pk] = edges[from_pk][to_pk] + amount if to_pk in edges[from_pk] else amount
```
Comment (vandenheuvel) [inline Tribler/community/multichain/community.py]: You could use the `+=` operator.

File: Tribler/community/multichain/community.py [hunk 2]
```diff
@@ -293,34 +293,97 @@ def get_statistics(self, public_key=None):
...
-        regarding the local multichain database.
+        regarding the local MultiChain database.
 
         :param public_key: the public key of the focus node in raw format
         :param neighbor_level: the radius within which the neighbors have to be returned
-        :return: a tuple of a dictionary with nodes and a dictionary with edges
+        :return: a tuple of a list with nodes and a list with edges
         """
         if public_key is None:
             public_key = self.my_member.public_key
```
Comment (vandenheuvel) [inline Tribler/community/multichain/community.py]: Instead of setting this optional argument here, you could do that in the function signature.

File: Tribler/community/multichain/community.py [hunk 3]
```diff
@@ -293,34 +293,97 @@ def get_statistics(self, public_key=None):
...
+
+        return_nodes = nodes.values()
+        return_edges = []
+        for node in nodes.keys():
+            self.page_rank.add_node(node)
+        for from_pk, edge_list in edges.iteritems():
+            new_edges = [{"from": from_pk, "to": to_pk, "amount": amount} for to_pk, amount in edge_list.iteritems()]
+            for edge in new_edges:
+                self.page_rank.add_edge(edge['from'], edge['to'], edge['amount'])
+            return_edges = return_edges + new_edges
```
Comment (vandenheuvel) [inline Tribler/community/multichain/community.py]: Once again, you could use the `+=` operator.

File: Tribler/community/multichain/community.py [hunk 4]
```diff
@@ -293,34 +293,97 @@ def get_statistics(self, public_key=None):
...
+
         self.get_page_ranks()
-        for dic in nodes:
+        for dic in return_nodes:
             dic["page_rank"] = 0
```
Comment (vandenheuvel) [inline Tribler/community/multichain/community.py]: You could write this as `dic["page_rank"] = self.ranks[dic["public_key"]] if dic["public_key"] in self.ranks else 0`.

File: Tribler/community/multichain/community.py [hunk 5]
```diff
@@ -293,34 +293,97 @@ def get_statistics(self, public_key=None):
...
+            self.page_rank.add_node(node)
+        for from_pk, edge_list in edges.iteritems():
+            new_edges = [{"from": from_pk, "to": to_pk, "amount": amount} for to_pk, amount in edge_list.iteritems()]
+            for edge in new_edges:
+                self.page_rank.add_edge(edge['from'], edge['to'], edge['amount'])
+            return_edges = return_edges + new_edges
+
         self.get_page_ranks()
-        for dic in nodes:
+        for dic in return_nodes:
```
Comment (vandenheuvel) [inline Tribler/community/multichain/community.py]: Could you use another variable name? How about `dictionary`, or `node`?

File: Tribler/community/multichain/database.py [hunk 6]
```diff
@@ -212,29 +212,43 @@ def total_traffic(self, public_key):
...
-        :param public_key: primary key of the focus node
-        :return: dictionary with for each neighbor of the focus a key, value entry: primary key neighbor, dictionary
-        containing the amount of data uploaded and downloaded from that neighbor
+        Get all edges for the graph focused on the given public key with the given neighbor level as depth.
+
+        The data will have the following format:
+        [(from_pk, to_pk, amount_up, amount_down, total_up, total_down), ...]
+
+        Please note that the total_up and total_down are from the node corresponding to the from_pk information. These
+        values might not be the most recent one, as it is taken from a block which might not be the last one. To get the
```
Comment (vandenheuvel) [inline Tribler/community/multichain/database.py]: The last one in the database or the last one created?

File: Tribler/community/multichain/database.py [hunk 7]
```diff
@@ -212,29 +212,43 @@ def total_traffic(self, public_key):
...
+        maximum total_up and total_down, make sure to take the maximum for a given node over different rows.
+
+        :param public_key: the public key which has to be focused
```
Comment (vandenheuvel) [inline Tribler/community/multichain/database.py]: "the public key of which the neighbors have to be retrieved".

File: Tribler/community/multichain/database.py [hunk 8]
```diff
@@ -212,29 +212,43 @@ def total_traffic(self, public_key):
...
-        for row in db_result:
-            neighbor_pk = row[0] if isinstance(row[0], str) else str(row[0])
-            neighbors[neighbor_pk] = {"up": row[1] or 0, "down": row[2] or 0}
-
-        return neighbors
+        # Select the first layer of neighbors.
+        query = u"""SELECT DISTINCT public_key, link_public_key FROM multi_chain
+                WHERE link_public_key = ? OR public_key = ?"""
+
+        # Recursively query the higher order neighbors.
```
Comment (vandenheuvel) [inline Tribler/community/multichain/database.py]: I suggest rewording this to something like "further away" neighbors.

File: Tribler/community/multichain/database.py [hunk 9]
```diff
@@ -212,29 +212,43 @@ def total_traffic(self, public_key):
...
+        for levelIndex in range(1, neighbor_level):
```
Comment (vandenheuvel) [inline Tribler/community/multichain/database.py]: You could just name the variable `level`.

File: Tribler/community/multichain/database.py [hunk 10]
```diff
@@ -212,29 +212,43 @@ def total_traffic(self, public_key):
...
+            query = u"""SELECT DISTINCT mc.public_key, mc.link_public_key FROM multi_chain mc,
+                    (%(lastlevel)s) level%(index)d
+                    WHERE level%(index)d.public_key = mc.public_key
+                    OR level%(index)d.public_key = mc.link_public_key
+                    OR level%(index)d.link_public_key = mc.public_key
+                    OR level%(index)d.link_public_key = mc.link_public_key
+                    """ % {"index": levelIndex, "lastlevel": query}
```
Comment (vandenheuvel) [inline Tribler/community/multichain/database.py]: You could rename this variable to `last_level`.

File: Tribler/community/multichain/community.py (hunk 2 above)
Comment (MaxVanDeursen) [inline Tribler/community/multichain/community.py]: This should not be done as self.my_member.public_key is subject to change. This change would then not be used in this function.
